pandas/core/arrays/datetimelike.py

```diff
@@ -366,15 +463,110 @@ def __getitem__(self, key):
 
         return self._simple_new(result, **attribs)
 
+    def __setitem__(
+            self,
+            key,    # type: Union[int, Sequence[int], Sequence[bool], slice]
+            value,  # type: Union[NaTType, Scalar, Sequence[Scalar]]
+    ):
+        # type: (...) -> None
+        # I'm fudging the types a bit here. The "Scalar" above really depends
+        # on type(self). For PeriodArray, it's Period (or stuff coercible
+        # to a period in from_sequence). For DatetimeArray, it's Timestamp...
+        # I don't know if mypy can do that, possibly with Generics.
+        # https://mypy.readthedocs.io/en/latest/generics.html
+
+        # n.b. This is moved from PeriodArray with the following changes
+        # 1. added is_slice check (bug on master)
+        # 2. changed dedicated ctor (period_array) to _from_sequence
+        # 3. Changed freq checking to use `_check_compatible_with`
+        # 4. Handle `value=iNaT` (may be able to revert. Check internals.)
+        if is_list_like(value):
+            is_slice = isinstance(key, slice)
+            if (not is_slice
+                    and len(key) != len(value)
+                    and not com.is_bool_indexer(key)):
+                msg = ("shape mismatch: value array of length '{}' does not "
+                       "match indexing result of length '{}'.")
+                raise ValueError(msg.format(len(key), len(value)))
+            if not is_slice and len(key) == 0:
+                return
+
+            value = type(self)._from_sequence(value, dtype=self.dtype)
```

Comment:
This is actually slightly more permissive that what's on master. On master, we didn't allow setting things coercible to Period

```python
In [13]: arr = pd.core.arrays.period_array([2000, 2001, 2002], freq='D')

In [14]: arr[[1]] = ["2000"]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
```

That's because we called `period_array` without passing `freq=self.freq`. Do we want to allow this?

Reply:
>  as officially, it should error on strings.

I don't think we officially state that. Just that *pandas* will call it with a sequence of scalars. But we're internally to the datetime modules using it, since we locally know that it handles more that just scalars.

Reply:
Yes, but personally, I would say that ideally we limit it to what it needs to be able to do. As otherwise it might mask places where we are doing something wrong (that eg might error with an external EA that does not do this), or we might inadvertently coerce back to DatetimeArray where the user actually wants strings (don't know if we would have actual cases of this though)